### PR TITLE
Update gen_protos to work with new protobuf on os-x.

### PR DESCRIPTION
**Pull request by robertwb**
Apparently the c-compiled protos are no longer available by default.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @AnandInguva

**Comment by robertwb** [comment]
The failures in apache_beam.runners.interactive.sql.utils_test.OptionsFormTest.test_dataflow_options_form look irrelevant.

**Comment by robertwb** [comment]
Run Python_Coverage PreCommit

**Comment by tvalentyn** [comment]
Run Python_Runners PreCommit

File: sdks/python/gen_protos.py [hunk 1]
```diff
@@ -123,7 +123,19 @@ def generate_urn_files(out_dir, api_path):
...
-  from google._upb import _message
+  try:
+    from google._upb import _message
+    list_types = (
+        list,
+        _message.RepeatedScalarContainer,
+        _message.RepeatedCompositeContainer,
+    )  # pylint: disable=c-extension-no-member
+  except ImportError:
+    from google.protobuf.internal import containers
```
Comment (tvalentyn) [inline sdks/python/gen_protos.py:134]: could you add a comment in which cases you expect else branch be used?

**Review by tvalentyn — APPROVED** [review]
asked someone to look into broken build.

File: sdks/python/gen_protos.py (hunk 1 above)
Comment (AnandInguva) [reply sdks/python/gen_protos.py:134]: Yes, that would be helpful.  I haven't caught this error while i was using when `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=upb`.
 but when `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION==python`(slow implementation of protobufs which we don't want to use)  I was using  `except` case.

Comment (robertwb) [reply sdks/python/gen_protos.py]: I've refactored this to be more explicit. Apparently the default implementation depends on the platform (and what wheels are shipped for it). Fast protos shouldn't matter here, and we are (as we should) verifying that we have a compiled implementation in our containers.

**Review by robertwb — COMMENTED** [review]
PTAL

**Comment by tvalentyn** [comment]
Backend: Wrote response {'code': 1, 'exc_type': 'AttributeError', 'exc_msg': "module 'google._upb._message' has no attribute 'RepeatedScalarFieldContainer'"} to /tmp/pep517_get_requires_for_build_sdist-xkfitq5_.json

**Comment by tvalentyn** [comment]
Would it be a problem if different implementations were used at pipeline submission vs runtime? Is this a scenario we may be facing (for example when one submits a pipeline from MacOS which presumably doesn't have the right wheels)?

File: sdks/python/gen_protos.py [hunk 2]
```diff
@@ -123,8 +123,21 @@ def generate_urn_files(out_dir, api_path):
...
+  elif api_implementation.Type() == 'upb':
+    from google._upb import _message as impl
+  elif api_implementation.Type() == 'cpp':
+    import google.protobuf.pyext._message as impl
+  else:
+    raise TypeError(
+        "Unknown proto implementation: " + api_implementation.Type())
+  repeated_types = (
+      list,
+      impl.RepeatedScalarFieldContainer,
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/gen_protos.py]: Oh, good point. Fixed.

**Comment by robertwb** [comment]
Regarding different implementations at compile vs. runtime, this is fine, as this code is just to do proto introspection to auto-generate the urn files to make type checkers more robust.

**Review by AnandInguva — APPROVED** [review]

**Review by AnandInguva — COMMENTED** [review]
LGTM. seems like there is a lint error.

**Comment by tvalentyn** [comment]
yes, I understand this code is not relevant for job submission. My question is whether we should anticipate issues for some users of Mac OS b/c they will be using a different protobuf implementation.

**Comment by robertwb** [comment]
Ah. I hope not; the various implementations (or at least their public APIs)
should be drop in replacements for each other. It's possible local runner
performance may suffer, but probably not by that much, and it was never
targeted at high efficiency anyway.

**Pull request merged**